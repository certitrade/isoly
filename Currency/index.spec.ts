import { isoly } from "../index"

describe("isoly.Currency", () => {
	it.each([
		[undefined, false],
		["SEK", true],
		["ZZZ", false],
		...isoly.Currency.values.map(v => [v, true] as const),
	] as const)("is(%s) == %s", (value, expected) => expect(isoly.Currency.is(value)).toEqual(expected))
	it.each([
		[0.004755520000000001, "SEK", 0],
		[0.004999920000000001, "SEK", 0],
		[17999.204999999998, "SEK", 17999.21],
		[0.005, "SEK", 0.01],
		[0.0049, "SEK", 0.0],
		[0.00499, "SEK", 0.0],
		[0.004999, "SEK", 0.0],
		[0.0049999, "SEK", 0.0],
		[0.00499999, "SEK", 0.0],
		[0.004999999, "SEK", 0.01],
		[0.004999981, "SEK", 0.0],
		[0.575, "EUR", 0.58],
		[0.855, "EUR", 0.86],
		[0.565, "EUR", 0.57],
		[0.055, "EUR", 0.06],
		[42.13377, "BHD", 42.134],
		[42.13377, "CLF", 42.1338],
		[42.13377, "EUR", 42.13],
		[42.13377, "ISK", 42],
		[42.13377, "XXX" as any, 42.13],
	] as const)("round(%f, %s) == %f", (value, currency, expected) =>
		expect(isoly.Currency.round(value, currency)).toEqual(expected)
	)
	it.each([
		[1234567890, "SEK", true],
		[10, "SEK", true],
		[1, "SEK", true],
		[1.1, "SEK", true],
		[1.11, "SEK", true],
		[1.111, "SEK", false],
		[1.1111, "SEK", false],
		[NaN, "SEK", false],
		[1234567890, "ISK", true],
		[10, "ISK", true],
		[1, "ISK", true],
		[1.1, "ISK", false],
		[1.11, "ISK", false],
		[1.12345678, "XAU", true],
		[12345678, "XAU", true],
	] as const)("isRounded(%f, %s) == %s", (value, currency, expected) =>
		expect(isoly.Currency.isRounded(value, currency)).toEqual(expected)
	)
	it.each([
		[95.385 + 18.114 + 0.001, 113.50000000000001],
		[isoly.Currency.add("ISK", 95.385 + 18.114, 0.001), 113],
		[isoly.Currency.add("SEK", 95.385 + 18.114, 0.001), 113.5],
		[isoly.Currency.add("BHD", 95.385 + 18.114, 0.001), 113.5],
		[4.3 + 0.1 + 0.1, 4.499999999999999],
		[isoly.Currency.add("ISK", 0.1, 4.3 + 0.1), 4],
		[isoly.Currency.add("SEK", 0.1, 4.3 + 0.1), 4.5],
		[isoly.Currency.add("BHD", 0.1, 4.3 + 0.1), 4.5],
	])("add(%f, %f) == %f", (value, expected) => expect(value).toEqual(expected))
	it.each([
		[2.1 / 0.6, 3.5000000000000004],
		[isoly.Currency.divide("ISK", 2.1, 0.6), 3],
		[isoly.Currency.divide("SEK", 2.1, 0.6), 3.5],
		[isoly.Currency.divide("BHD", 2.1, 0.6), 3.5],
	])("divide(%f, %f) == %f", (value, expected) => expect(value).toEqual(expected))
	it.each([
		["ISK", 2090.5, 8.61, 18004],
		["SEK", 2090.5, 8.61, 17999.21],
		["BHD", 2090.5, 8.61, 17999.205],
	] as const)("multiply(%s, %f, %f) == %f", (currency, left, right, expected) =>
		expect(isoly.Currency.multiply(currency, left, right)).toEqual(expected)
	)
	it.each([
		[2090.5, 8.61, 17999.204999999998],
		[1000, 50000 * 0.000001, 49.99999999999999],
		[10, 0.000001 * 50000, 0.49999999999999994], //should be 0.5
		[0.000001, 50000, 0.049999999999999996], //should be 0.05
		[10, 0.1 * 1.5, 1.5000000000000002], //should be 1.5
		[0.1, 1.5, 0.15000000000000002], //should be 0.15
	])("%f * %f == %f", (left, right, expected) => expect(left * right).toEqual(expected))
	it.each([
		["ISK", 2090.5, 8.61, 18004],
		["SEK", 2090.5, 8.61, 17999.21],
		["BHD", 2090.5, 8.61, 17999.205],
		["ISK", 1000, 0.000001 * 50000, 50],
		["SEK", 1000, 0.000001 * 50000, 50],
		["BHD", 1000, 0.000001 * 50000, 50],
		["ISK", 10, 0.000001 * 50000, 1],
		["SEK", 10, 0.000001 * 50000, 0.5],
		["BHD", 10, 0.1 * 1.5, 1.5],
		["ISK", 0.1, 1.5, 0],
		["SEK", 0.1, 1.5, 0.15],
		["BHD", 0.1, 1.5, 0.15],
	] as const)("multiply %s, %f, %f, %f == %f", (currency, left, right, expected) =>
		expect(isoly.Currency.multiply(currency, left, right)).toEqual(expected)
	)
	it.each([
		["SEK", 0.1, 1.5, "ISK", 0],
		["SEK", 0.1, 1.5, "SEK", 0.15],
		["SEK", 0.1, 1.5, "BHD", 0.15],
	] as const)("convert %s, %f, %f, %s == %f", (currencyLeft, left, right, currencyRight, expected) =>
		expect(isoly.Currency.convert(currencyLeft, left, right, currencyRight)).toEqual(expected)
	)
	it.each([
		[1.4, 0.9, 0.4999999999999999], // should be 0.5
		[16.4, 0.9, 15.499999999999998], // should be 15.5
		[76.9, 21.4, 55.50000000000001], // should be 55.5
	])("%f - %f == %f", (left, right, expected) => expect(left - right).toEqual(expected))
	it.each([
		["ISK", 16.4, 0.9, 15],
		["ISK", 1.4, 0.9, 0],
		["SEK", 1.4, 0.9, 0.5],
		["BHD", 1.4, 0.9, 0.5],
		["SEK", 16.4, 0.9, 15.5],
		["BHD", 16.4, 0.9, 15.5],
		["ISK", 76.9, 21.4, 56],
		["SEK", 76.9, 21.4, 55.5],
		["BHD", 76.9, 21.4, 55.5],
	] as const)("subtract %s, %f, %f == %f", (currency, left, right, expected) =>
		expect(isoly.Currency.subtract(currency, left, right)).toEqual(expected)
	)
	it.each([
		[10.42, 3.01, 0, 0, 13.43],
		[10.4, 3.03, 0, 0, 13.43],
		[10.2, 3.03, 0.2, 0, 13.43],
		[9.81, 3.41, 0.21, 0, 13.43],
		[10.02, 3.2, 0.21, 0, 13.43],
		[5.01, 5.19, 3.03, 0.2, 13.43],
		[0.11, 0.11, 0.11, 0, 0.33],
		[0.01, 0.08, 0.24, 0, 0.33],
		[0.13, 0.13, 0.11, 0, 0.37],
		[0.11, 0.09, 0.14, 0, 0.34],
		[0, 0.01, 0.56, 0.14, 0.71],
		[-0.01, 0.02, 0.56, 0.14, 0.71],
		[-0.01, 0.02, 0.55, 0.14, 0.7],
	])("sum %j == %f", (term0, term1, term2, term3, expected) =>
		expect(
			isoly.Currency.add("EUR", isoly.Currency.add("EUR", isoly.Currency.add("EUR", term0, term1), term2), term3)
		).toEqual(expected)
	)
	it.each([
		[0.004755520000000001, "SEK", 0],
		[0.004999920000000001, "SEK", 0],
		[0, "SEK", 0],
		[0.01, "SEK", 1],
		[0.02, "SEK", 2],
		[0.049, "SEK", 5],
		[0.0501, "SEK", 5],
		[3.05, "SEK", 305],
		[0.5, "SEK", 50],
		[0.2, "ISK", 0],
		[400.2, "ISK", 400],
		[400, "ISK", 400],
		[123.456, "XPT", 123.456],
	] as const)("toMinor %f, %s == %f", (value, currency, expected) =>
		expect(isoly.Currency.toMinor(value, currency)).toEqual(expected)
	)
	it.each([
		[0, "SEK", 0],
		[1, "SEK", 0.01],
		[2, "SEK", 0.02],
		[5, "SEK", 0.05],
		[305, "SEK", 3.05],
		[50, "SEK", 0.5],
		[0, "ISK", 0],
		[400, "ISK", 400],
		[123.456, "XPT", 123.456],
	] as const)("fromMinor %f, %s == %f", (value, currency, expected) =>
		expect(isoly.Currency.fromMinor(value, currency)).toEqual(expected)
	)
	it.each(isoly.Currency.values)("name(%s)", currency => expect(isoly.Currency.name(currency)).toMatchSnapshot())
})
