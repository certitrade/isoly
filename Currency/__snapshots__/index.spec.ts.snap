// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Currency > name(AED) 1`] = `"United Arab Emirates dirham"`;

exports[`Currency > name(AFN) 1`] = `"Afghan afghani"`;

exports[`Currency > name(ALL) 1`] = `"Albanian lek"`;

exports[`Currency > name(AMD) 1`] = `"Armenian dram"`;

exports[`Currency > name(ANG) 1`] = `"Netherlands Antillean guilder"`;

exports[`Currency > name(AOA) 1`] = `"Angolan kwanza"`;

exports[`Currency > name(ARS) 1`] = `"Argentine peso"`;

exports[`Currency > name(AUD) 1`] = `"Australian dollar"`;

exports[`Currency > name(AWG) 1`] = `"Aruban florin"`;

exports[`Currency > name(AZN) 1`] = `"Azerbaijani manat"`;

exports[`Currency > name(BAM) 1`] = `"Bosnia and Herzegovina convertible mark"`;

exports[`Currency > name(BBD) 1`] = `"Barbados dollar"`;

exports[`Currency > name(BDT) 1`] = `"Bangladeshi taka"`;

exports[`Currency > name(BGN) 1`] = `"Bulgarian lev"`;

exports[`Currency > name(BHD) 1`] = `"Bahraini dinar"`;

exports[`Currency > name(BIF) 1`] = `"Burundian franc"`;

exports[`Currency > name(BMD) 1`] = `"Bermudian dollar"`;

exports[`Currency > name(BND) 1`] = `"Brunei dollar"`;

exports[`Currency > name(BOB) 1`] = `"Boliviano"`;

exports[`Currency > name(BOV) 1`] = `"Bolivian Mvdol"`;

exports[`Currency > name(BRL) 1`] = `"Brazilian real"`;

exports[`Currency > name(BSD) 1`] = `"Bahamian dollar"`;

exports[`Currency > name(BTN) 1`] = `"Bhutanese ngultrum"`;

exports[`Currency > name(BWP) 1`] = `"Botswana pula"`;

exports[`Currency > name(BYN) 1`] = `"Belarusian ruble"`;

exports[`Currency > name(BZD) 1`] = `"Belize dollar"`;

exports[`Currency > name(CAD) 1`] = `"Canadian dollar"`;

exports[`Currency > name(CDF) 1`] = `"Congolese franc"`;

exports[`Currency > name(CHE) 1`] = `"WIR euro"`;

exports[`Currency > name(CHF) 1`] = `"Swiss franc"`;

exports[`Currency > name(CHW) 1`] = `"WIR franc"`;

exports[`Currency > name(CLF) 1`] = `"Unidad de Fomento"`;

exports[`Currency > name(CLP) 1`] = `"Chilean peso"`;

exports[`Currency > name(CNY) 1`] = `"Renminbi"`;

exports[`Currency > name(COP) 1`] = `"Colombian peso"`;

exports[`Currency > name(COU) 1`] = `"Unidad de Valor Real"`;

exports[`Currency > name(CRC) 1`] = `"Costa Rican colon"`;

exports[`Currency > name(CUC) 1`] = `"peso cubano convertible"`;

exports[`Currency > name(CUP) 1`] = `"Cuban peso"`;

exports[`Currency > name(CVE) 1`] = `"Cape Verdean escudo"`;

exports[`Currency > name(CZK) 1`] = `"Czech koruna"`;

exports[`Currency > name(DJF) 1`] = `"Djiboutian franc"`;

exports[`Currency > name(DKK) 1`] = `"Danish krone"`;

exports[`Currency > name(DOP) 1`] = `"Dominican peso"`;

exports[`Currency > name(DZD) 1`] = `"Algerian dinar"`;

exports[`Currency > name(EGP) 1`] = `"Egyptian pound"`;

exports[`Currency > name(ERN) 1`] = `"Eritrean nakfa"`;

exports[`Currency > name(ETB) 1`] = `"Ethiopian birr"`;

exports[`Currency > name(EUR) 1`] = `"Euro"`;

exports[`Currency > name(FJD) 1`] = `"Fiji dollar"`;

exports[`Currency > name(FKP) 1`] = `"Falkland Islands pound"`;

exports[`Currency > name(GBP) 1`] = `"Pound sterling"`;

exports[`Currency > name(GEL) 1`] = `"Georgian lari"`;

exports[`Currency > name(GHS) 1`] = `"Ghanaian cedi"`;

exports[`Currency > name(GIP) 1`] = `"Gibraltar pound"`;

exports[`Currency > name(GMD) 1`] = `"Gambian dalasi"`;

exports[`Currency > name(GNF) 1`] = `"Guinean franc"`;

exports[`Currency > name(GTQ) 1`] = `"Guatemalan quetzal"`;

exports[`Currency > name(GYD) 1`] = `"Guyanese dollar"`;

exports[`Currency > name(HKD) 1`] = `"Hong Kong dollar"`;

exports[`Currency > name(HNL) 1`] = `"Honduran lempira"`;

exports[`Currency > name(HRK) 1`] = `"Hrvatska kuna"`;

exports[`Currency > name(HTG) 1`] = `"Haitian gourde"`;

exports[`Currency > name(HUF) 1`] = `"Hungarian forint"`;

exports[`Currency > name(IDR) 1`] = `"Indonesian rupiah"`;

exports[`Currency > name(ILS) 1`] = `"Israeli new shekel"`;

exports[`Currency > name(INR) 1`] = `"Indian rupee"`;

exports[`Currency > name(IQD) 1`] = `"Iraqi dinar"`;

exports[`Currency > name(IRR) 1`] = `"Iranian rial"`;

exports[`Currency > name(ISK) 1`] = `"Icelandic króna"`;

exports[`Currency > name(JMD) 1`] = `"Jamaican dollar"`;

exports[`Currency > name(JOD) 1`] = `"Jordanian dinar"`;

exports[`Currency > name(JPY) 1`] = `"Japanese yen"`;

exports[`Currency > name(KES) 1`] = `"Kenyan shilling"`;

exports[`Currency > name(KGS) 1`] = `"Kyrgyzstani som"`;

exports[`Currency > name(KHR) 1`] = `"Cambodian riel"`;

exports[`Currency > name(KMF) 1`] = `"Comoro franc"`;

exports[`Currency > name(KPW) 1`] = `"North Korean won"`;

exports[`Currency > name(KRW) 1`] = `"South Korean won"`;

exports[`Currency > name(KWD) 1`] = `"Kuwaiti dinar"`;

exports[`Currency > name(KYD) 1`] = `"Cayman Islands dollar"`;

exports[`Currency > name(KZT) 1`] = `"Kazakhstani tenge"`;

exports[`Currency > name(LAK) 1`] = `"Lao kip"`;

exports[`Currency > name(LBP) 1`] = `"Lebanese pound"`;

exports[`Currency > name(LKR) 1`] = `"Sri Lankan rupee"`;

exports[`Currency > name(LRD) 1`] = `"Liberian dollar"`;

exports[`Currency > name(LSL) 1`] = `"Lesotho loti"`;

exports[`Currency > name(LYD) 1`] = `"Libyan dinar"`;

exports[`Currency > name(MAD) 1`] = `"Moroccan dirham"`;

exports[`Currency > name(MDL) 1`] = `"Moldovan leu"`;

exports[`Currency > name(MGA) 1`] = `"Malagasy ariary"`;

exports[`Currency > name(MKD) 1`] = `"Macedonian denar"`;

exports[`Currency > name(MMK) 1`] = `"Myanmar kyat"`;

exports[`Currency > name(MNT) 1`] = `"Mongolian tögrög"`;

exports[`Currency > name(MOP) 1`] = `"Macanese pataca"`;

exports[`Currency > name(MRU) 1`] = `"Mauritanian ouguiya"`;

exports[`Currency > name(MUR) 1`] = `"Mauritian rupee"`;

exports[`Currency > name(MVR) 1`] = `"Maldivian rufiyaa"`;

exports[`Currency > name(MWK) 1`] = `"Malawian kwacha"`;

exports[`Currency > name(MXN) 1`] = `"Mexican peso"`;

exports[`Currency > name(MXV) 1`] = `"Mexican Unidad de Inversion"`;

exports[`Currency > name(MYR) 1`] = `"Malaysian ringgit"`;

exports[`Currency > name(MZN) 1`] = `"Mozambican metical"`;

exports[`Currency > name(NAD) 1`] = `"Namibian dollar"`;

exports[`Currency > name(NGN) 1`] = `"Nigerian naira"`;

exports[`Currency > name(NIO) 1`] = `"Nicaraguan córdoba"`;

exports[`Currency > name(NOK) 1`] = `"Norwegian krone"`;

exports[`Currency > name(NPR) 1`] = `"Nepalese rupee"`;

exports[`Currency > name(NZD) 1`] = `"New Zealand dollar"`;

exports[`Currency > name(OMR) 1`] = `"Omani rial"`;

exports[`Currency > name(PAB) 1`] = `"Panamanian balboa"`;

exports[`Currency > name(PEN) 1`] = `"Peruvian sol"`;

exports[`Currency > name(PGK) 1`] = `"Papua New Guinean kina"`;

exports[`Currency > name(PHP) 1`] = `"Philippine peso"`;

exports[`Currency > name(PKR) 1`] = `"Pakistani rupee"`;

exports[`Currency > name(PLN) 1`] = `"Polish złoty"`;

exports[`Currency > name(PYG) 1`] = `"Paraguayan guaraní"`;

exports[`Currency > name(QAR) 1`] = `"Qatari riyal"`;

exports[`Currency > name(RON) 1`] = `"Romanian leu"`;

exports[`Currency > name(RSD) 1`] = `"Serbian dinar"`;

exports[`Currency > name(RUB) 1`] = `"Russian ruble"`;

exports[`Currency > name(RWF) 1`] = `"Rwandan franc"`;

exports[`Currency > name(SAR) 1`] = `"Saudi riyal"`;

exports[`Currency > name(SBD) 1`] = `"Solomon Islands dollar"`;

exports[`Currency > name(SCR) 1`] = `"Seychelles rupee"`;

exports[`Currency > name(SDG) 1`] = `"Sudanese pound"`;

exports[`Currency > name(SEK) 1`] = `"Swedish krona"`;

exports[`Currency > name(SGD) 1`] = `"Singapore dollar"`;

exports[`Currency > name(SHP) 1`] = `"Saint Helena pound"`;

exports[`Currency > name(SLL) 1`] = `"Sierra Leonean leone"`;

exports[`Currency > name(SOS) 1`] = `"Somalian shilling"`;

exports[`Currency > name(SRD) 1`] = `"Surinamese dollar"`;

exports[`Currency > name(SSP) 1`] = `"South Sudanese pound"`;

exports[`Currency > name(STN) 1`] = `"São Tomé and Príncipe dobra"`;

exports[`Currency > name(SVC) 1`] = `"Salvadoran colón"`;

exports[`Currency > name(SYP) 1`] = `"Syrian pound"`;

exports[`Currency > name(SZL) 1`] = `"Swazi lilangeni"`;

exports[`Currency > name(THB) 1`] = `"Thai baht"`;

exports[`Currency > name(TJS) 1`] = `"Tajikistani somoni"`;

exports[`Currency > name(TMT) 1`] = `"Turkmenistan manat"`;

exports[`Currency > name(TND) 1`] = `"Tunisian dinar"`;

exports[`Currency > name(TOP) 1`] = `"Tongan paʻanga"`;

exports[`Currency > name(TRY) 1`] = `"Turkish lira"`;

exports[`Currency > name(TTD) 1`] = `"Trinidad and Tobago dollar"`;

exports[`Currency > name(TWD) 1`] = `"New Taiwan dollar"`;

exports[`Currency > name(TZS) 1`] = `"Tanzanian shilling"`;

exports[`Currency > name(UAH) 1`] = `"Ukrainian hryvnia"`;

exports[`Currency > name(UGX) 1`] = `"Ugandan shilling"`;

exports[`Currency > name(USD) 1`] = `"United States dollar"`;

exports[`Currency > name(USN) 1`] = `"United States dollar"`;

exports[`Currency > name(UYI) 1`] = `"Uruguay Peso en Unidades Indexadas"`;

exports[`Currency > name(UYU) 1`] = `"Uruguayan peso"`;

exports[`Currency > name(UYW) 1`] = `"Unidad previsional"`;

exports[`Currency > name(UZS) 1`] = `"Uzbekistani sum"`;

exports[`Currency > name(VES) 1`] = `"Venezuelan sovereign bolívar"`;

exports[`Currency > name(VND) 1`] = `"Vietnamese đồng"`;

exports[`Currency > name(VUV) 1`] = `"Vanuatu vatu"`;

exports[`Currency > name(WST) 1`] = `"Samoan tala"`;

exports[`Currency > name(XAF) 1`] = `"CFA franc BEAC"`;

exports[`Currency > name(XAG) 1`] = `"Silver"`;

exports[`Currency > name(XAU) 1`] = `"Gold"`;

exports[`Currency > name(XBA) 1`] = `"European Composite Unit"`;

exports[`Currency > name(XBB) 1`] = `"European Monetary Unit"`;

exports[`Currency > name(XBC) 1`] = `"European Unit of Account 9"`;

exports[`Currency > name(XBD) 1`] = `"European Unit of Account 17"`;

exports[`Currency > name(XCD) 1`] = `"East Caribbean dollar"`;

exports[`Currency > name(XDR) 1`] = `"Special drawing rights"`;

exports[`Currency > name(XOF) 1`] = `"CFA franc BCEAO"`;

exports[`Currency > name(XPD) 1`] = `"Palladium"`;

exports[`Currency > name(XPF) 1`] = `"CFP franc"`;

exports[`Currency > name(XPT) 1`] = `"Platinum"`;

exports[`Currency > name(XSU) 1`] = `"SUCRE"`;

exports[`Currency > name(XTS) 1`] = `"Code reserved for testing"`;

exports[`Currency > name(XUA) 1`] = `"ADB Unit of Account"`;

exports[`Currency > name(XXX) 1`] = `"No currency"`;

exports[`Currency > name(YER) 1`] = `"Yemeni rial"`;

exports[`Currency > name(ZAR) 1`] = `"South African rand"`;

exports[`Currency > name(ZMW) 1`] = `"Zambian kwacha"`;

exports[`Currency > name(ZWL) 1`] = `"Zimbabwean dollar"`;
